src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java

```diff
@@ -0,0 +1,959 @@
+            LOGGER.error("Received a modded custom payload packet from a server that is not in the configuration or play phase. Disconnecting server.");
+            throw new IllegalStateException("A server send a packet while not in the configuration or play phase. Somebody changed the phases known to NeoForge!");
+        }
+        return true;
+    }
+
+    /**
+     * Invoked by the server when it completes the negotiation with the client during the configuration phase.
+     * <p>
+     * This method determines what the versions of each of the channels are, and checks if the client and server have a compatible set of network channels.
+     * </p>
+     * <p>
+     * If the negotiation fails, a custom packet is send to the client to inform it of the failure, and which will allow the client to disconnect gracefully with an indicative error screen.
+     * </p>
+     * <p>
+     * This method should only be invoked for modded connections. Use {@link #onVanillaConnectionDetectedAtServer(ServerConfigurationPacketListener)} to indicate that during the configuration phase of the network handshake between a client and the server, a vanilla connection was detected.
+     * </p>
+     *
+     * @param sender        The listener which completed the negotiation.
+     * @param configuration The configuration channels that the client has available.
+     * @param play          The play channels that the client has available.
+     */
+    public void onModdedConnectionDetectedAtServer(ServerConfigurationPacketListener sender, Set<ModdedNetworkQueryComponent> configuration, Set<ModdedNetworkQueryComponent> play) {
+        final NegotiationResult configurationNegotiationResult = NetworkComponentNegotiator.negotiate(
+                knownConfigurationRegistrations.entrySet().stream()
+                        .map(entry -> new NegotiableNetworkComponent(entry.getKey(), entry.getValue().version(), entry.getValue().flow(), entry.getValue().optional()))
+                        .toList(),
+                configuration.stream()
+                        .map(entry -> new NegotiableNetworkComponent(entry.id(), entry.version(), entry.flow(), entry.optional()))
+                        .toList());
+
+        sender.getConnection().channel().attr(ATTRIBUTE_IS_MODDED_CONNECTION).set(true);
+        sender.getConnection().channel().attr(ATTRIBUTE_FLOW).set(PacketFlow.SERVERBOUND);
+        sender.getConnection().channel().attr(ATTRIBUTE_PAYLOAD_SETUP).set(NetworkPayloadSetup.emptyModded());
+
+        //Negotiation failed. Disconnect the client.
+        if (!configurationNegotiationResult.success()) {
+            if (!configurationNegotiationResult.failureReasons().isEmpty()) {
+                sender.send(new ModdedNetworkSetupFailedPayload(configurationNegotiationResult.failureReasons()));
+            }
+
+            sender.disconnect(Component.translatable("multiplayer.disconnect.incompatible", "NeoForge %s".formatted(NeoForgeVersion.getVersion())));
```

Comment:
Is there meant to be a return statement after this line so that it doesn't check the negotiation of play? Or is it meant to potentially have multiple `sender.disconnect` calls?